graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -1,35 +1,148 @@
 # Copyright 2018-present Kensho Technologies, LLC.
 
+import six
+
+from .constants import UNSUPPORTED_META_FIELDS
+from .sql_tree import SqlNode, SqlQueryTree
+from .. import blocks
+from ..ir_lowering_sql.constants import SqlOutput
+from ..metadata import LocationInfo
+from ... import exceptions
+
 ##############
 # Public API #
 ##############
 
+_SKIPPABLE_BLOCKS = (
+    # MarkLocation blocks are used in the first pass over the IR blocks to create a mapping of
+    # IR block -> query path for all IR blocks. They can safely be skipped during tree construction.
+    blocks.MarkLocation,
+)
+
 
 def lower_ir(ir_blocks, query_metadata_table, type_equivalence_hints=None):
     """Lower the IR into a form that can be represented by a SQL query.
 
     Args:
-        ir_blocks: list of IR blocks to lower into SQL-compatible form
+        ir_blocks: list of IR blocks to lower into the SQL tree structure.
         query_metadata_table: QueryMetadataTable object containing all metadata collected during
                               query processing, including location metadata (e.g. which locations
                               are folded or optional).
-        type_equivalence_hints: optional dict of GraphQL interface or type -> GraphQL union.
-                                Used as a workaround for GraphQL's lack of support for
-                                inheritance across "types" (i.e. non-interfaces), as well as a
-                                workaround for Gremlin's total lack of inheritance-awareness.
-                                The key-value pairs in the dict specify that the "key" type
-                                is equivalent to the "value" type, i.e. that the GraphQL type or
-                                interface in the key is the most-derived common supertype
-                                of every GraphQL type in the "value" GraphQL union.
-                                Recursive expansion of type equivalence hints is not performed,
-                                and only type-level correctness of this argument is enforced.
-                                See README.md for more details on everything this parameter does.
-                                *****
-                                Be very careful with this option, as bad input here will
-                                lead to incorrect output queries being generated.
-                                *****
+        type_equivalence_hints: optional dict of GraphQL interface or type -> GraphQL union. Unused.
 
     Returns:
         tree representation of IR blocks for recursive traversal by SQL backend.
     """
-    raise NotImplementedError(u'SQL IR lowering is not yet implemented.')
+    construct_result = ir_blocks.pop()
```

Comment:
Assert that this is indeed the `ConstructResult` block. If that ever changes, this implicit assumption will be a nightmare to track down.

Reply:
This assumption is handled in the validation pass now.

Reply:
I wouldn't `pop()` here. The compiler generally tries to avoid mutating existing data structures (especially ones provided by the caller) since that tends to prevent a few different classes of bugs. Can you just select the block you need instead?

Also, I wouldn't put that assertion in the validation pass. When looking at this code, it's not obvious that such an assertion exists (so this code still raises red flags for me), and when looking at the validation pass, it's not clear what the assertion is meant to guard.

Consider making a helper function that in one go asserts what you need and then returns the data you need.